src/inference/ess.jl

```diff
@@ -0,0 +1,134 @@
+"""
+    ESS
+
+Elliptical slice sampling algorithm.
+
+# Examples
+```jldoctest; setup = :(Random.seed!(1))
+julia> @model gdemo(x) = begin
+           m ~ Normal()
+           x ~ Normal(m, 0.5)
+       end
+gdemo (generic function with 2 methods)
+
+julia> sample(gdemo(1.0), ESS(), 1_000) |> mean
+Mean
+
+│ Row │ parameters │ mean     │
+│     │ Symbol     │ Float64  │
+├─────┼────────────┼──────────┤
+│ 1   │ m          │ 0.824853 │
+```
+"""
+struct ESS{space} <: InferenceAlgorithm end
+
+ESS() = ESS{()}()
+ESS(space::Symbol) = ESS{(space,)}()
+
+mutable struct ESSState{V<:VarInfo} <: AbstractSamplerState
+    vi::V
+end
+
+function Sampler(alg::ESS, model::Model, s::Selector)
+    # sanity check
+    vi = VarInfo(model)
+    space = getspace(alg)
+    vns = _getvns(vi, s, Val(space))
+    length(vns) == 1 ||
+        error("[ESS] does only support one variable ($(length(vns)) variables specified)")
+    dist = getdist(vi, vns[1][1])
+    isgaussian(dist) ||
+        error("[ESS] only supports Gaussian prior distributions")
+
+    state = ESSState(vi)
+    info = Dict{Symbol, Any}()
+
+    return Sampler(alg, info, s, state)
+end
+
+isgaussian(dist) = false
+isgaussian(::Normal) = true
+isgaussian(::NormalCanon) = true
+isgaussian(::AbstractMvNormal) = true
+
+# always accept in the first step
+function step!(::AbstractRNG, model::Model, spl::Sampler{<:ESS}, ::Integer; kwargs...)
+    return Transition(spl)
+end
+
+function step!(
+    rng::AbstractRNG,
+    model::Model,
+    spl::Sampler{<:ESS},
+    ::Integer,
+    ::Transition;
+    kwargs...
+)
+    # obtain mean of distribution
+    vi = spl.state.vi
+    vn = _getvns(vi, spl)[1][1]
+    dist = getdist(vi, vn)
+    μ = vectorize(dist, mean(dist))
+
+    # obtain previous sample
+    f = vi[vn]
+
+    # recompute log-likelihood in logp
+    if spl.selector.tag !== :default
+        runmodel!(model, vi, spl)
+    end
+    setgid!(vi, spl.selector, vn)
+
+    # sample log-likelihood threshold for the next sample
+    threshold = getlogp(vi) - randexp(rng)
+
+    # sample from the prior
+    ν = vectorize(dist, rand(rng, dist))
+
+    # sample initial angle
+    θ = 2 * π * rand(rng)
+    θₘᵢₙ = θ - 2 * π
```

Comment:
I would suggest to not use too many Unicode characters. It makes maintaining the code more difficult and doesn't seem necessary here.

Reply:
I just tried to use the notation in the original paper to make it easier to understand, and assumed that it's easier to read with Unicode characters. I don't have a strong opinion on this, so I'll change it to `theta`, `thetamin`, `pi` etc.

Reply:
Thanks. We had a lot of Unicode in the old NUTS sampler for the same reasons. This turned out to be harder to maintain than using clear variable names. It’s often easier to understand what is happening if the variable names are telling, which is not necessarily the case for the notation used in the paper.

Reply:
BTW: I’m fine with the \theta but I referred to the variable you assign a value to.